File: granhotel/backend/app/models/reservation.py
import enum
from sqlalchemy import Column, Integer, String, Boolean, DateTime, func, Enum as SAEnum, ForeignKey, Numeric, Date
from sqlalchemy.orm import relationship
from ..db.base_class import Base

class ReservationStatus(str, enum.Enum):
    PENDING = "PENDING"
    CONFIRMED = "CONFIRMED"
    CHECKED_IN = "CHECKED_IN"
    CHECKED_OUT = "CHECKED_OUT"
    CANCELLED = "CANCELLED"
    NO_SHOW = "NO_SHOW"
    WAITLIST = "WAITLIST" # As per issue spec for waitlist management

class Reservation(Base):
    __tablename__ = "reservations"

    id = Column(Integer, primary_key=True, index=True)

    guest_id = Column(Integer, ForeignKey("guests.id"), nullable=False, index=True)
    room_id = Column(Integer, ForeignKey("rooms.id"), nullable=False, index=True) # Assuming direct room assignment for now

    check_in_date = Column(Date, nullable=False)
    check_out_date = Column(Date, nullable=False)
    reservation_date = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    status = Column(SAEnum(ReservationStatus, name="reservation_status_enum"), nullable=False, default=ReservationStatus.PENDING, index=True)

    # total_price will store the calculated price. Numeric for precision.
    # Precision and scale can be adjusted based on currency needs.
    total_price = Column(Numeric(10, 2), nullable=True) # Allow null initially if price calculated later

    notes = Column(String, nullable=True)

    # Placeholder for who booked it, assuming a User model will exist later
    # booked_by_user_id = Column(Integer, ForeignKey("users.id"), nullable=True)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    # Relationships
    guest = relationship("Guest", backref="reservations") # Simple backref for now
    room = relationship("Room", backref="reservations")   # Simple backref

    # user = relationship("User", backref="reservations") # For booked_by_user_id

File: granhotel/frontend/src/types/reservation.ts
// src/types/reservation.ts
import { Guest } from './guest';
import { Room } from './room';

// Matches backend models.reservation.ReservationStatus enum
export enum ReservationStatus {
  PENDING = "PENDING",
  CONFIRMED = "CONFIRMED",
  CHECKED_IN = "CHECKED_IN",
  CHECKED_OUT = "CHECKED_OUT",
  CANCELLED = "CANCELLED",
  NO_SHOW = "NO_SHOW",
  WAITLIST = "WAITLIST",
}

export interface Reservation {
  id: number;
  guest_id: string; // UUID string from backend Guest model
  room_id: number;
  check_in_date: string; // ISO date string (e.g., "YYYY-MM-DD")
  check_out_date: string; // ISO date string
  reservation_date: string; // ISO datetime string
  status: ReservationStatus;
  total_price?: number | null; // Backend sends Decimal, frontend might receive as number.
  notes?: string | null;
  created_at: string; // ISO datetime string
  updated_at: string; // ISO datetime string

  guest?: Guest | null;
  room?: Room | null;
}

export interface ReservationCreatePayload {
  guest_id: string; // UUID string
  room_id: number;
  check_in_date: string; // "YYYY-MM-DD"
  check_out_date: string; // "YYYY-MM-DD"
  status?: ReservationStatus;
  notes?: string | null;
}

export interface ReservationUpdatePayload {
  guest_id?: string;
  room_id?: number;
  check_in_date?: string; // "YYYY-MM-DD"
  check_out_date?: string; // "YYYY-MM-DD"
  status?: ReservationStatus;
  total_price?: number | null;
  notes?: string | null;
}

// For updating status specifically using the PATCH endpoint
// The actual payload for the PATCH request might be query parameters, not a body.
// This schema is for type safety of the status value itself.
export interface ReservationStatusUpdatePayload {
    new_status: ReservationStatus;
}
